### PR TITLE
Implement framework-specific AdversarialPatch attack for TensorFlow v2

### LINKED ISSUE #347: Implement framework-specific version of Adversarial Patch attack - TensorFlow v2
Implement framework-specific version of Adversarial Patch attack for TensorFlow v2.

The new implementation of `AdversarialPatch` should use the new framework-specific implementation created with this issue if a classifier of type `TensorFlowV2Classifier` is provided and use the current framework-independent implementation for all other classifier types.

Paper link: https://arxiv.org/abs/1712.09665

### PR DESCRIPTION
**Pull request by beat-buesser**
# Description

This pull request adds the first framework-specific attack implementation in ART for the example of `AdversarialPatch` attack in TensorFlow v2. The framework-specific implementations will be used where available, otherwise the framework-independent implementation will be used. Framework-specific implementations will allow ART to take advantage full advantage of each framework especially in terms of speed and accuracy.

Fixes #347 

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/attacks/test_adversarial_patch.py [hunk 1]
```diff
@@ -99,14 +115,17 @@ def test_pytorch(self):
...
         attack_ap = AdversarialPatch(
-            ptc, rotation_max=22.5, scale_min=0.1, scale_max=1.0, learning_rate=5.0, batch_size=10, max_iter=500
+            ptc, rotation_max=0.5, scale_min=0.4, scale_max=0.41, learning_rate=5.0, batch_size=10, max_iter=500
         )
+        master_seed(seed=1234)
+        target = np.zeros(self.x_train_mnist.shape[0])
+        patch_adv, _ = attack_ap.generate(x_train, target)
 
-        patch_adv, _ = attack_ap.generate(x_train)
+        # patch_adv = np.reshape(patch_adv, (28, 28, 1)).astype(np.float32)
```
Comment (hesseltuinhof) [inline tests/attacks/test_adversarial_patch.py:124]: If commenting out is desired, can we also remove the line?

File: art/attacks/evasion/adversarial_patch/adversarial_patch_numpy.py [hunk 2]
```diff
@@ -38,43 +37,37 @@
...
         rotation_max=22.5,
         scale_min=0.1,
         scale_max=1.0,
         learning_rate=5.0,
         max_iter=500,
-        clip_patch=None,
         batch_size=16,
     ):
         """
         Create an instance of the :class:`.AdversarialPatch`.
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/adversarial_patch/adversarial_patch_numpy.py:67]: Yes, I agree

File: art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+        rotation_max=22.5,
+        scale_min=0.1,
+        scale_max=1.0,
+        learning_rate=5.0,
+        max_iter=500,
+        batch_size=16,
+        patch_shape=None,
+    ):
+        """
+        Create an instance of the :class:`.AdversarialPatch`.
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py:68]: Yes, I agree

File: .travis.yml [hunk 4]
```diff
@@ -15,8 +15,8 @@ matrix:
         - py.test --pep8 -m pep8
 
 before_install:
-  - sed -i "s/tensorflow/tensorflow==${TENSORFLOW_V}/" test_requirements.txt
-  - sed -i "s/keras/keras==${KERAS_V}/" test_requirements.txt
+  - sed -i "s/tensorflow\b/tensorflow==${TENSORFLOW_V}/" test_requirements.txt
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply .travis.yml:18]: That's a great idea.

File: art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+            "max_iter": max_iter,
+            "batch_size": batch_size,
+            "patch_shape": patch_shape,
+        }
+        self.set_params(**kwargs)
+
+        self.image_shape = classifier.input_shape
+
+        assert self.image_shape[2] in [1, 3], "Color channel need to be in last dimension"
+        assert self.patch_shape[2] in [1, 3], "Color channel need to be in last dimension"
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/adversarial_patch/adversarial_patch_tensorflow.py:112]: The patch_shape can be different form image_shape for the framework-specific implementation.

File: art/attacks/evasion/adversarial_patch/adversarial_patch.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+This module implements the adversarial patch attack `AdversarialPatch`. This attack generates an adversarial patch that
+can be printed into the physical world with a common printer. The patch can be used to fool image classifiers.
+
+| Paper link: https://arxiv.org/abs/1712.09665
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+
+import numpy as np
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/adversarial_patch/adversarial_patch.py:28]: I agree

File: art/attacks/evasion/adversarial_patch/adversarial_patch_numpy.py [hunk 7]
```diff
@@ -162,22 +148,29 @@ def generate(self, x, y=None, **kwargs):
                     )
                     patch_gradients += patch_gradients_i
 
-            patch_gradients = patch_gradients / (num_batches * self.batch_size)
+            # patch_gradients = patch_gradients / (num_batches * self.batch_size)
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/evasion/adversarial_patch/adversarial_patch_numpy.py:151]: Good catch, but I would like to keep this line for historical reasons.

File: tests/attacks/test_adversarial_patch.py (hunk 1 above)
Comment (beat-buesser) [reply tests/attacks/test_adversarial_patch.py:124]: I agree
